IntelliJ IDEA enables nightly feature of thiserror on stable compiler

## Environment

* **IntelliJ Rust plugin version:** 0.4.185.5064-223-nightly
* **Rust toolchain version:** 1.65.0 (897e37553 2022-11-02) x86_64-pc-windows-msvc
* **IDE name and version:** IntelliJ IDEA 2022.3 Ultimate Edition (IU-223.7571.182)
* **Operating system:** Windows 10 10.0
* **Macro expansion:** enabled
* **Additional experimental features:** org.rust.cargo.emulate.terminal

## Problem description
When adding `thiserror` to a project, IntelliJ IDEA somehow enables the `provide_any` nightly feature in it, causing compilation to fail until `cargo clean` is run.

## Steps to reproduce
1. Create a new binary project
2. Add `thiserror = "1.0.37"` to its dependencies and click "Load Cargo Changes"
3. Once this is done and indexing is complete, run the project.

The result is compilation failing with

```
C:/Users/L/.cargo/bin/cargo.exe build --color=always --message-format=json-diagnostic-rendered-ansi --package idebug --bin idebug
   Compiling thiserror v1.0.37
error[E0554]: `#![feature]` may not be used on the stable release channel
   --> C:\Users\L\.cargo\registry\src\github.com-1ecc6299db9ec823\thiserror-1.0.37\src\lib.rs:238:34
    |
238 | #![cfg_attr(provide_any, feature(provide_any))]
    |                                  ^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0554`.
error: could not compile `thiserror` due to 2 previous errors
Process finished with exit code 101
```

Performing `cargo clean` fixes this, allowing compilation to succeed. The project can then print "Hello, world!".

VSCode with rust-analyzer apparently hit this same issue https://github.com/dtolnay/thiserror/issues/192 which was fixed in an update at some point.

--- a/native-helper/src/rustc_wrapper.rs
+++ b/native-helper/src/rustc_wrapper.rs
@@ -33,11 +33,29 @@ pub fn run_rustc_skipping_cargo_checking(
 }
 
 fn run_rustc(rustc_executable: OsString, args: Vec<OsString>) -> io::Result<ExitCode> {
-    let mut child = Command::new(rustc_executable)
+    const RUSTC_BOOTSTRAP: &'static str = "RUSTC_BOOTSTRAP";
+    const ORIGINAL_RUSTC_BOOTSTRAP: &'static str = "INTELLIJ_ORIGINAL_RUSTC_BOOTSTRAP";
+
+    let mut command = Command::new(rustc_executable);
+    command
         .args(args)
         .stdin(Stdio::inherit())
         .stdout(Stdio::inherit())
-        .stderr(Stdio::inherit())
+        .stderr(Stdio::inherit());
+
+    // IntelliJ Rust plugin may add `RUSTC_BOOTSTRAP=1` environment variable to `cargo check` call
+    // during build script execution.
+    // But we need it only to use experimental `cargo` feature
+    // and don't want to propagate it to `rustc` call because it may produce different results.
+    // So we use `INTELLIJ_ORIGINAL_RUSTC_BOOTSTRAP` environment variable to keep original value
+    // of `RUSTC_BOOTSTRAP` and restore it (if needed) for `rustc` call.
+    //
+    // See https://github.com/intellij-rust/intellij-rust/issues/9700
+    command.env_remove(RUSTC_BOOTSTRAP);
+    if let Some(original_rustc_bootstrap_value) = std::env::var_os(ORIGINAL_RUSTC_BOOTSTRAP) {
+        command.env(RUSTC_BOOTSTRAP, original_rustc_bootstrap_value);
+    }
+    let mut child = command
         .spawn()?;
     Ok(ExitCode(child.wait()?.code()))
 }

--- a/src/main/kotlin/org/rust/cargo/toolchain/RsToolchainBase.kt
+++ b/src/main/kotlin/org/rust/cargo/toolchain/RsToolchainBase.kt
@@ -127,6 +127,12 @@ abstract class RsToolchainBase(val location: Path) {
         const val RUSTC_BOOTSTRAP: String = "RUSTC_BOOTSTRAP"
         const val RUSTC_WRAPPER: String = "RUSTC_WRAPPER"
 
+        /**
+         * Environment variable used to keep original value of `RUSTC_BOOTSTRAP` variable
+         * to be able to restore original value for subprocesses if needed
+         */
+        const val ORIGINAL_RUSTC_BOOTSTRAP = "INTELLIJ_ORIGINAL_RUSTC_BOOTSTRAP"
+
         @JvmOverloads
         fun suggest(projectDir: Path? = null): RsToolchainBase? {
             val distribution = projectDir?.let { WslPath.getDistributionByWindowsUncPath(it.toString()) }

--- a/src/main/kotlin/org/rust/cargo/toolchain/tools/Cargo.kt
+++ b/src/main/kotlin/org/rust/cargo/toolchain/tools/Cargo.kt
@@ -45,6 +45,7 @@ import org.rust.cargo.runconfig.command.CargoCommandConfiguration.Companion.find
 import org.rust.cargo.runconfig.command.CargoCommandConfiguration.Companion.findCargoTargets
 import org.rust.cargo.runconfig.command.workingDirectory
 import org.rust.cargo.toolchain.*
+import org.rust.cargo.toolchain.RsToolchainBase.Companion.ORIGINAL_RUSTC_BOOTSTRAP
 import org.rust.cargo.toolchain.RsToolchainBase.Companion.RUSTC_BOOTSTRAP
 import org.rust.cargo.toolchain.RsToolchainBase.Companion.RUSTC_WRAPPER
 import org.rust.cargo.toolchain.impl.BuildMessages
@@ -292,7 +293,17 @@ class Cargo(
         val envMap = mutableMapOf<String, String>()
         if (useKeepGoing) {
             additionalArgs += listOf("-Z", "unstable-options", "--keep-going")
+            val originalRustcBootstrapValue = System.getenv(RUSTC_BOOTSTRAP)
             envMap += RUSTC_BOOTSTRAP to "1"
+            // Actually, we need to pass `RUSTC_BOOTSTRAP=1` environment variable here only for `cargo`
+            // and don't want to propagate it to `rustc` call because it may produce different results.
+            // So we use `INTELLIJ_ORIGINAL_RUSTC_BOOTSTRAP` environment variable to keep original value
+            // of `RUSTC_BOOTSTRAP` and restore it (if needed) for `rustc` call in native-helper binary
+            //
+            // See https://github.com/intellij-rust/intellij-rust/issues/9700
+            if (originalRustcBootstrapValue != null) {
+                envMap += ORIGINAL_RUSTC_BOOTSTRAP to originalRustcBootstrapValue
+            }
         }
         val nativeHelper = RsPathManager.nativeHelper(toolchain is RsWslToolchain)
         if (nativeHelper != null && USE_BUILD_SCRIPT_WRAPPER.asBoolean()) {

--- a/src/test/kotlin/org/rust/cargo/project/model/impl/SyncToolWindowTest.kt
+++ b/src/test/kotlin/org/rust/cargo/project/model/impl/SyncToolWindowTest.kt
@@ -442,6 +442,44 @@ class SyncToolWindowTest : RsWithToolchainTestBase() {
         """)
     }
 
+    @WithExperimentalFeatures(EVALUATE_BUILD_SCRIPTS)
+    fun `test build script evaluation`() {
+        val project = buildProject {
+            toml("Cargo.toml", """
+                [package]
+                name = "hello"
+                version = "0.1.0"
+                authors = []
+            """)
+
+            dir("src") {
+                rust("main.rs", """
+                    fn main() {}
+                """)
+            }
+            rust("build.rs", """
+                const RUSTC_BOOTSTRAP: Option<&str> = option_env!("RUSTC_BOOTSTRAP");
+
+                fn main() {
+                    if let Some(value) = RUSTC_BOOTSTRAP {
+                        if value == "1" {
+                            panic!("Unexpected `RUSTC_BOOTSTRAP=1` env during compilation")
+                        }
+                    }
+                }
+            """)
+        }
+        checkSyncViewTree("""
+            -
+             -finished
+              -Sync ${project.root.name} project
+               Getting toolchain version
+               -Updating workspace info
+                -Build scripts evaluation
+                 Compiling hello v0.1.0
+               Getting Rust stdlib
+        """)
+    }
 
     private fun attachCargoProject(cargoProjectRoot: VirtualFile) {
         myFixture.launchAction("Cargo.AttachCargoProject", PlatformDataKeys.VIRTUAL_FILE to cargoProjectRoot)

